File: src/modules/user/user.model.ts
import mongoose, { Schema } from "mongoose";
import {
  IUser,
  IOTP,
  IUserMemory,
  ISessionSummary,
  IWorkoutHistory,
} from "./user.interface";

// ─────────────────────────────────────────────────────────────
// SUB-SCHEMAS
// ─────────────────────────────────────────────────────────────

const sessionSummarySchema = new Schema<ISessionSummary>(
  {
    date: { type: Date, default: Date.now },
    workoutSummary: { type: String },
    exercisesCompleted: [{ type: String }],
    loadsUsed: { type: Schema.Types.Mixed, default: {} },
    adherence: {
      type: String,
      enum: ["completed", "skipped", "modified"],
      default: "completed",
    },
    rpe: { type: Number, min: 1, max: 10 },
    painNotes: { type: String },
    sleepStressRating: { type: Number, min: 1, max: 10 },
    energyLevel: { type: Number, min: 1, max: 10 },
    nextSessionPrescribed: { type: String },
    flags: [{ type: String }],
  },
  { _id: true },
);

const userMemorySchema = new Schema<IUserMemory>(
  {
    trainerId: {
      type: Schema.Types.ObjectId,
      ref: "Trainer",
      required: true,
    },
    profileMemory: {
      preferredName: { type: String },
      goal: {
        type: String,
        enum: ["maintain_physique", "muscle_gain", "weight_loss", "boxing"],
      },
      experienceLevel: {
        type: String,
        enum: ["beginner", "intermediate", "advanced"],
      },
      scheduleDaysPerWeek: { type: Number },
      equipment: { type: String },
      limitations: { type: String },
      preferences: { type: String },
      motivationStyle: {
        type: String,
        enum: ["tough_love", "gentle", "balanced"],
        default: "balanced",
      },
      updatedAt: { type: Date, default: Date.now },
    },
    rollingMemory: {
      last3Sessions: [sessionSummarySchema],
      lastKnownLoads: { type: Schema.Types.Mixed, default: {} },
      adherenceNotes: { type: String },
      recoveryNotes: { type: String },
      flags: [{ type: String }],
      updatedAt: { type: Date, default: Date.now },
    },
    lastUpdatedAt: { type: Date, default: Date.now },
  },
  { _id: true },
);

const workoutHistorySchema = new Schema<IWorkoutHistory>(
  {
    trainerId: { type: Schema.Types.ObjectId, ref: "Trainer" },
    date: { type: Date, default: Date.now },
    focus: { type: String },
    exercisesPerformed: [
      {
        exerciseId: { type: Schema.Types.ObjectId },
        exerciseName: { type: String },
        blockName: { type: String },
        sets: { type: Number },
        reps: { type: String },
        weight: { type: String },
        rpe: { type: Number },
        completed: { type: Boolean, default: true },
      },
    ],
    sessionRpe: { type: Number, min: 1, max: 10 },
    durationMinutes: { type: Number },
    notes: { type: String },
    aiPlanUsed: { type: Boolean, default: false },
  },
  { _id: true },
);

// ─────────────────────────────────────────────────────────────
// MAIN USER SCHEMA
// ─────────────────────────────────────────────────────────────

const userSchema = new Schema<IUser>(
  {
    // ── Existing fields (unchanged) ───────────────────────────
    firstName: { type: String, required: true },
    lastName: { type: String, required: true },
    email: { type: String, required: true, unique: true, index: true },
    dateOfBirth: { type: String, required: false },
    password: { type: String, required: true },
    gender: {
      type: String,
      enum: ["male", "female", "not_prefer_to_say"],
      required: true,
    },
    role: {
      type: String,
      enum: ["admin", "user", "trainer"],
      default: "user",
    },
    profilePicture: { type: String, required: false },
    coverPhoto: { type: String, required: false },
    bio: { type: String, required: false },
    isVerified: { type: Boolean, required: true, default: false },
    isDeleted: { type: Boolean, required: true, default: false },
    fcmToken: { type: String, required: false },

    // ── Fitness Profile ──────────────────────────────────────
    height: { type: Number, required: false },
    weight: { type: Number, required: false },
    fitnessLevel: {
      type: String,
      enum: ["beginner", "intermediate", "advanced"],
      required: false,
    },
    injuries: [{ type: String }],
    availableEquipment: { type: String },
    trainingDaysPerWeek: { type: Number, min: 1, max: 7 },

    // ── Goal ─────────────────────────────────────────────────
    primaryGoal: {
      type: String,
      enum: ["maintain_physique", "muscle_gain", "weight_loss", "boxing"],
      required: false,
    },

    // ── Trainer Subscription ──────────────────────────────────
    subscribedTrainer: {
      type: Schema.Types.ObjectId,
      ref: "Trainer",
      required: false,
    },
    subscriptionTier: {
      type: String,
      enum: ["free", "paid", "premium", "monthly", "annual"],
      default: "free",
    },
    subscriptionStartDate: { type: Date, required: false },
    subscriptionEndDate: { type: Date, required: false },

    // ── Anam AI Usage (new) ───────────────────────────────────
    // Tracks monthly video call minutes per user
    // Auto-initialized on first call, resets every 30 days
    anamAI: {
      monthlyMinutesLimit: {
        type: Number,
        default: 250, // platform default: 250 min/month
      },
      minutesUsedThisMonth: {
        type: Number,
        default: 0,
      },
      currentPeriodStart: {
        type: Date,
        default: Date.now, // set when user makes first call
      },
      totalMinutesAllTime: {
        type: Number,
        default: 0,
      },
    },

    // ── AI Memory ────────────────────────────────────────────
    memory: [userMemorySchema],

    // ── Workout History ──────────────────────────────────────
    workoutHistory: [workoutHistorySchema],

    // ── Onboarding ───────────────────────────────────────────
    onboardingCompleted: { type: Boolean, default: false },
  },
  { timestamps: true },
);

// ─────────────────────────────────────────────────────────────
// INSTANCE METHODS
// ─────────────────────────────────────────────────────────────

userSchema.methods.getMemoryForTrainer = function (trainerId: string) {
  return this.memory.find(
    (m: IUserMemory) => m.trainerId.toString() === trainerId.toString(),
  );
};

// ─────────────────────────────────────────────────────────────
// INDEXES
// ─────────────────────────────────────────────────────────────

userSchema.index({ subscribedTrainer: 1 });
userSchema.index({ primaryGoal: 1 });
userSchema.index({ "memory.trainerId": 1 });

// ─────────────────────────────────────────────────────────────
// MODELS
// ─────────────────────────────────────────────────────────────

export const UserModel = mongoose.model<IUser>("User", userSchema);

// ─────────────────────────────────────────────────────────────
// OTP (unchanged)
// ─────────────────────────────────────────────────────────────

const OTPSchema = new Schema<IOTP>({
  email: { type: String, required: true, trim: true, index: true },
  otp: { type: String, required: true, trim: true },
  expiresAt: { type: Date, required: true, index: { expires: "1m" } },
});

export const OTPModel = mongoose.model<IOTP>("OTP", OTPSchema);
OTPSchema.index({ email: 1, expiresAt: 1 });

File: lib/features/profile/data/models/user_model.dart
import 'package:intl/intl.dart';

class UserModel {
  String? sId;
  String? firstName;
  String? lastName;
  String? email;
  String? gender;
  String? role;
  bool? isVerified;
  bool? isDeleted;
  List<String>? injuries;
  String? subscriptionTier;
  bool? onboardingCompleted;

  List<Memory>? memory;
  List<dynamic>? workoutHistory;

  String? createdAt;
  String? updatedAt;
  int? iV;

  AnamAI? anamAI;

  String? availableEquipment;
  String? dateOfBirth;
  String? fitnessLevel;
  int? height;
  String? primaryGoal;
  int? trainingDaysPerWeek;
  num? weight;
  String? bio;

  String? subscriptionStartDate;
  String? subscriptionEndDate;
  String? profilePicture;
  String? coverPhoto;
  String? preferredName;
  String? motivationStyle;

  SubscribedTrainer? subscribedTrainer;

  UserModel({
    this.sId,
    this.firstName,
    this.lastName,
    this.email,
    this.gender,
    this.role,
    this.isVerified,
    this.isDeleted,
    this.injuries,
    this.subscriptionTier,
    this.onboardingCompleted,
    this.memory,
    this.workoutHistory,
    this.createdAt,
    this.updatedAt,
    this.iV,
    this.anamAI,
    this.availableEquipment,
    this.dateOfBirth,
    this.fitnessLevel,
    this.height,
    this.primaryGoal,
    this.trainingDaysPerWeek,
    this.weight,
    this.bio,
    this.subscriptionStartDate,
    this.subscriptionEndDate,
    this.profilePicture,
    this.coverPhoto,
    this.preferredName,
    this.motivationStyle,
    this.subscribedTrainer,
  });

  String get fullName => "$firstName $lastName";

  String get subscriptionPeriod {
    if (subscriptionStartDate == null || subscriptionEndDate == null) {
      return 'N/A';
    }
    final start = DateTime.parse(subscriptionStartDate!).toLocal();
    final end = DateTime.parse(subscriptionEndDate!).toLocal();
    final formatter = DateFormat('d MMMM yyyy');
    return '${formatter.format(start)} - ${formatter.format(end)}';
  }

  UserModel.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    firstName = json['firstName'];
    lastName = json['lastName'];
    email = json['email'];
    gender = json['gender'];
    role = json['role'];
    isVerified = json['isVerified'];
    isDeleted = json['isDeleted'];

    injuries = json['injuries'] is List
        ? (json['injuries'] as List).map((e) => e.toString()).toList()
        : [];

    subscriptionTier = json['subscriptionTier'];
    onboardingCompleted = json['onboardingCompleted'];

    memory = json['memory'] is List
        ? (json['memory'] as List)
            .whereType<Map>()
            .map((e) => Memory.fromJson(Map<String, dynamic>.from(e)))
            .toList()
        : [];

    workoutHistory = json['workoutHistory'] is List
        ? List<dynamic>.from(json['workoutHistory'] as List)
        : [];

    createdAt = json['createdAt'];
    updatedAt = json['updatedAt'];
    iV = json['__v'];

    anamAI = json['anamAI'] is Map
        ? AnamAI.fromJson(Map<String, dynamic>.from(json['anamAI'] as Map))
        : null;

    availableEquipment = json['availableEquipment'];
    dateOfBirth = json['dateOfBirth'];
    fitnessLevel = json['fitnessLevel'];
    height = json['height'];
    primaryGoal = json['primaryGoal'];
    trainingDaysPerWeek = json['trainingDaysPerWeek'];
    weight = json['weight'] != null ? num.tryParse(json['weight'].toString()) : null;
    bio = json['bio'];

    subscriptionStartDate = json['subscriptionStartDate'];
    subscriptionEndDate = json['subscriptionEndDate'];
    profilePicture = json['profilePicture'];
    coverPhoto = json['coverPhoto'];
    preferredName = json['preferredName'];
    motivationStyle = json['motivationStyle'];

    subscribedTrainer = json['subscribedTrainer'] is Map
        ? SubscribedTrainer.fromJson(
            Map<String, dynamic>.from(json['subscribedTrainer'] as Map),
          )
        : null;
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': sId,
      'firstName': firstName,
      'lastName': lastName,
      'email': email,
      'gender': gender,
      'role': role,
      'isVerified': isVerified,
      'isDeleted': isDeleted,
      'injuries': injuries,
      'subscriptionTier': subscriptionTier,
      'onboardingCompleted': onboardingCompleted,
      'memory': memory?.map((e) => e.toJson()).toList(),
      'workoutHistory': workoutHistory,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
      '__v': iV,
      'anamAI': anamAI?.toJson(),
      'availableEquipment': availableEquipment,
      'dateOfBirth': dateOfBirth,
      'fitnessLevel': fitnessLevel,
      'height': height,
      'primaryGoal': primaryGoal,
      'trainingDaysPerWeek': trainingDaysPerWeek,
      'weight': weight,
      'bio': bio,
      'subscriptionStartDate': subscriptionStartDate,
      'subscriptionEndDate': subscriptionEndDate,
      'profilePicture': profilePicture,
      'coverPhoto': coverPhoto,
      'preferredName': preferredName,
      'motivationStyle': motivationStyle,
      'subscribedTrainer': subscribedTrainer?.toJson(),
    };
  }
}

class AnamAI {
  String? currentPeriodStart;
  int? minutesUsedThisMonth;
  int? monthlyMinutesLimit;
  int? totalMinutesAllTime;

  AnamAI({
    this.currentPeriodStart,
    this.minutesUsedThisMonth,
    this.monthlyMinutesLimit,
    this.totalMinutesAllTime,
  });

  AnamAI.fromJson(Map<String, dynamic> json) {
    currentPeriodStart = json['currentPeriodStart'];
    minutesUsedThisMonth = json['minutesUsedThisMonth'];
    monthlyMinutesLimit = json['monthlyMinutesLimit'];
    totalMinutesAllTime = json['totalMinutesAllTime'];
  }

  Map<String, dynamic> toJson() {
    return {
      'currentPeriodStart': currentPeriodStart,
      'minutesUsedThisMonth': minutesUsedThisMonth,
      'monthlyMinutesLimit': monthlyMinutesLimit,
      'totalMinutesAllTime': totalMinutesAllTime,
    };
  }
}

class Memory {
  String? trainerId;
  ProfileMemory? profileMemory;
  RollingMemory? rollingMemory;
  String? lastUpdatedAt;
  String? sId;

  Memory({
    this.trainerId,
    this.profileMemory,
    this.rollingMemory,
    this.lastUpdatedAt,
    this.sId,
  });

  Memory.fromJson(Map<String, dynamic> json) {
    trainerId = json['trainerId']?.toString();
    profileMemory = json['profileMemory'] is Map
        ? ProfileMemory.fromJson(
            Map<String, dynamic>.from(json['profileMemory'] as Map),
          )
        : null;

    rollingMemory = json['rollingMemory'] is Map
        ? RollingMemory.fromJson(
            Map<String, dynamic>.from(json['rollingMemory'] as Map),
          )
        : null;

    lastUpdatedAt = json['lastUpdatedAt'];
    sId = json['_id'];
  }

  Map<String, dynamic> toJson() {
    return {
      'trainerId': trainerId,
      'profileMemory': profileMemory?.toJson(),
      'rollingMemory': rollingMemory?.toJson(),
      'lastUpdatedAt': lastUpdatedAt,
      '_id': sId,
    };
  }
}

class ProfileMemory {
  String? preferredName;
  String? goal;
  String? experienceLevel;
  int? scheduleDaysPerWeek;
  String? equipment;
  String? limitations;
  String? preferences;
  String? updatedAt;
  String? motivationStyle;

  ProfileMemory({
    this.preferredName,
    this.goal,
    this.experienceLevel,
    this.scheduleDaysPerWeek,
    this.equipment,
    this.limitations,
    this.preferences,
    this.updatedAt,
    this.motivationStyle,
  });

  ProfileMemory.fromJson(Map<String, dynamic> json) {
    preferredName = json['preferredName'];
    goal = json['goal'];
    experienceLevel = json['experienceLevel'];
    scheduleDaysPerWeek = json['scheduleDaysPerWeek'];
    equipment = json['equipment'];
    limitations = json['limitations'];
    preferences = json['preferences'];
    updatedAt = json['updatedAt'];
    motivationStyle = json['motivationStyle'];
  }

  Map<String, dynamic> toJson() {
    return {
      'preferredName': preferredName,
      'goal': goal,
      'experienceLevel': experienceLevel,
      'scheduleDaysPerWeek': scheduleDaysPerWeek,
      'equipment': equipment,
      'limitations': limitations,
      'preferences': preferences,
      'updatedAt': updatedAt,
      'motivationStyle': motivationStyle,
    };
  }
}

class RollingMemory {
  List<dynamic>? last3Sessions;
  String? adherenceNotes;
  String? recoveryNotes;
  List<dynamic>? flags;
  String? updatedAt;

  RollingMemory({
    this.last3Sessions,
    this.adherenceNotes,
    this.recoveryNotes,
    this.flags,
    this.updatedAt,
  });

  RollingMemory.fromJson(Map<String, dynamic> json) {
    last3Sessions = json['last3Sessions'] != null
        ? List<dynamic>.from(json['last3Sessions'])
        : [];

    adherenceNotes = json['adherenceNotes'];
    recoveryNotes = json['recoveryNotes'];

    flags = json['flags'] != null
        ? List<dynamic>.from(json['flags'])
        : [];

    updatedAt = json['updatedAt'];
  }

  Map<String, dynamic> toJson() {
    return {
      'last3Sessions': last3Sessions,
      'adherenceNotes': adherenceNotes,
      'recoveryNotes': recoveryNotes,
      'flags': flags,
      'updatedAt': updatedAt,
    };
  }
}

class SubscribedTrainer {
  String? sId;
  String? name;
  List<String>? certifications;
  String? specialty;
  List<String>? trainingStyleTags;

  SubscribedTrainer({
    this.sId,
    this.name,
    this.certifications,
    this.specialty,
    this.trainingStyleTags,
  });

  SubscribedTrainer.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    name = json['name'];

    certifications = json['certifications'] != null
        ? List<String>.from(json['certifications'])
        : [];

    specialty = json['specialty'];

    trainingStyleTags = json['trainingStyleTags'] != null
        ? List<String>.from(json['trainingStyleTags'])
        : [];
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': sId,
      'name': name,
      'certifications': certifications,
      'specialty': specialty,
      'trainingStyleTags': trainingStyleTags,
    };
  }
}